mypy/main.py

```diff
@@ -280,83 +280,56 @@ def parse_version(v):
-    """
-    assert os.path.isdir(arg), arg
-    arg = os.path.abspath(arg)
-    for ext in PY_EXTENSIONS:
-        f = os.path.join(arg, '__init__' + ext)
-        if os.path.exists(f):
-            top_dir = crawl_up(f)[0]
-            break
-    else:
-        top_dir = arg
-
-    mods = set()  # type: Set[str]
+def expand_dir(arg: str, mod_prefix: str = '') -> List[BuildSource]:
+    """Convert a directory name to a list of sources to build."""
+    f = has_init_file(arg)
+    if mod_prefix and not f:
+        return []
+    seen = set()  # type: Set[str]
     sources = []
-    for root, dirs, files in os.walk(arg):
-        # Sort dirs and files in place to create a stable order.
-        dirs.sort()
-        files.sort()
-        # Look for dirs that are packages.
-        packages = []
-        for pkg in dirs:
-            d = os.path.join(root, pkg)
-            for ext in PY_EXTENSIONS:
-                f = os.path.join(d, '__init__' + ext)
-                mod = make_mod(top_dir, f, ext)
-                if os.path.exists(f):
-                    if mod not in mods:
-                        sources.append(BuildSource(f, mod, None))
-                        mods.add(mod)
-                        packages.append(pkg)
+    if f and not mod_prefix:
+        top_dir, top_mod = crawl_up(f)
+        mod_prefix = top_mod + '.'
+    if mod_prefix:
+        sources.append(BuildSource(f, mod_prefix.rstrip('.'), None))
+    names = os.listdir(arg)
+    names.sort(key=keyfunc)
+    for name in names:
+        path = os.path.join(arg, name)
+        if os.path.isdir(path):
+            sub_sources = expand_dir(path, mod_prefix + name + '.')
+            if sub_sources:
+                seen.add(name)
+                sources.extend(sub_sources)
+        else:
+            base, suffix = os.path.splitext(name)
+            if base == '__init__':
+                continue
+            if base not in seen and '.' not in base:
+                for ext in PY_EXTENSIONS:
+                    if suffix == ext:
+                        seen.add(base)
+                        src = BuildSource(path, mod_prefix + base, None)
+                        sources.append(src)
                         break
-        # Only descend into packages.
-        dirs[:] = packages
-        # Add modules at this level.  Note that (like Python) if
-        # both foo.py and foo/__init__.py exist, the package wins.
-        for ext in PY_EXTENSIONS:
-            # Note that .pyi gets preferred over .py because it
-            # comes first in PY_EXTENSIONS.
-            for f in files:
-                f = os.path.join(root, f)
-                if f.endswith(ext):
-                    mod = make_mod(top_dir, f, ext)
-                    if mod not in mods:
-                        sources.append(BuildSource(f, mod, None))
-                        mods.add(mod)
-    if not sources:
-        fail("There are no .py[i] files in directory '{}'".format(arg))
     return sources
 
 
```

Comment:
`make_mod` (defined on the next line, but github won't let me comment there) appears to no longer be used.